### ISSUE TITLE
Parametric constructor not called after adding it

### ISSUE THREAD

**Issue by mauro3** [JuliaLang/julia#23216]
This works fine (on 0.6 and 0.7):
```

julia> struct ST{T} end                                                    

julia> ST{T}(::T) where T = "$T"     
                                                                                           
julia> ST{Int}(5)                                                  
"Int64"                                                                                                                                                
```
but if it gets called once before adding that constructor, then it can't find it after either:
```
julia> struct ST{T} end                                                    

julia> ST{Int}(5)                    
ERROR: MethodError: Cannot `convert` an object of type Int64 to an object of type ST{Int64}
This may have arisen from a call to the constructor ST{Int64}(...),
since type constructors fall back to convert methods.                                                                                                  
Stacktrace:                                                                
 [1] ST{Int64}(::Int64) at ./sysimg.jl:24                                  
                                          
julia> ST{T}(::T) where T = "$T"     
                                                                 
julia> ST{Int}(5)                                                  
ERROR: MethodError: Cannot `convert` an object of type Int64 to an object of type ST{Int64}
This may have arisen from a call to the constructor ST{Int64}(...),
since type constructors fall back to convert methods.                  
Stacktrace:                                                    
 [1] ST{Int64}(::Int64) at ./sysimg.jl:24    
```
Note that this bug is not present with non-parametric types.

**Comment by StefanKarpinski** [comment]
@vtjnash – missing backedge?

**Comment by vtjnash** [comment]
It's a missing unionall allocation in typemap from the introduction of unionall (previously, the code assumed that `!(S <: T.parameters[1])` implies `!(V{S} <: T{S})`, but now, it needs to allocate unionall bounds to get subtyping to answer the same question)

**Comment by sostock** [comment]
This has been fixed since Julia 0.7.
